packages/core/src/AlertBar/AlertBar.js

```diff
@@ -20,110 +20,133 @@ import { Message } from './Message.js'
  * @see Specification: {@link https://github.com/dhis2/design-system/blob/master/molecules/alertbar.md|Design system}
  * @see Live demo: {@link /demo/?path=/story/alertbar--default|Storybook}
  */
-class AlertBar extends Component {
-    // visible is used to control the show/hide animation
-    // hidden is used to stop rendering entirely after hide animation
-    state = {
-        visible: true,
-        hidden: false,
+const AlertBar = ({
+    actions,
+    children,
+    className,
+    critical,
+    dataTest,
+    duration,
+    hidden,
+    icon,
+    permanent,
+    success,
+    warning,
+    onHidden,
+}) => {
+    const [inViewport, setInViewport] = useState(() => !hidden)
+    const [inDOM, setInDOM] = useState(() => !hidden)
+    const showTimeout = useRef(null)
+    const displayTimeout = useRef(null)
+    const hideTimeout = useRef(null)
+    const displayStartTime = useRef(null)
+    const displayTimeRemaining = useRef(null)
+    const info = !critical && !success && !warning
+    const shouldAutoHide = !(permanent || warning || critical)
+    const show = () => {
+        setInDOM(true)
+        setInViewport(true)
     }
-
-    componentDidMount() {
-        this.startTime = Date.now()
-        this.timeRemaining = this.props.duration
-        this.startDisplayTimeout()
+    const hide = () => {
+        setInDOM(true)
+        setInViewport(false)
     }
-
-    componentWillUnmount() {
-        clearTimeout(this.displayTimeout)
-        clearTimeout(this.onHiddenTimeout)
+    const remove = () => {
+        setInDOM(false)
+        setInViewport(false)
+        onHidden && onHidden({}, null)
     }
-
-    startDisplayTimeout = () => {
-        if (this.shouldAutoHide()) {
-            this.displayTimeout = setTimeout(() => {
-                this.hide(null)
-            }, this.timeRemaining)
+    const clearAllTimeouts = () => {
+        clearTimeout(showTimeout.current)
+        clearTimeout(displayTimeout.current)
+        clearTimeout(hideTimeout.current)
+    }
+    const runHideAnimation = () => {
+        clearAllTimeouts()
+        hide()
+        hideTimeout.current = setTimeout(remove, ANIMATION_TIME)
```

Comment:
I know you mentioned previously that you were looking into using the `animationend` event instead of using `setTimeout` - was `setTimeout` used due to browser support issues (and so we could look into using `animationend` in the future) or was this the simpler implementation?

Reply:
It was decided to not use the `animationend` event because critical functionality depends on it, and support is not rock-solid. If the event would not fire the `onHidden` callback would not be called and this could lead to serious issues. For example the DOM could get cluttered with lots of alert-bars by the alert service. See [this Slack thread](https://dhis2.slack.com/archives/CBM8LNEQM/p1622472566054000) to read how the decision came about.

So we decided to play it safe and stick with the `setTimeout` approach. We've got three timed events, the show-animation, the hide-animation and the display duration. If I had gone with the `animationend` event based implementation I would have been able to replace the timeouts for the show/hide animations with an event-listener and I would expect that to be a lot easier to reason about.

Having said that, I think the current implementation is actually not too complex now that I have switched to more explicit variable names, and opted for a "brute force" approach to clearing timeouts (I just clear all of them all the time to make things simpler).